pkg/database/vercode.go

```diff
@@ -97,11 +96,51 @@ func (v *VerificationCode) AfterCreate(scope *gorm.Scope) {
 				SET codes_issued = authorized_app_stats.codes_issued + 1
 		`
 
-		day := time.Now().UTC().Truncate(24 * time.Hour)
-		if err := scope.DB().Exec(sql, day, v.IssuingAppID).Error; err != nil {
+		if err := scope.DB().Exec(sql, v.CreatedAt, v.IssuingAppID).Error; err != nil {
 			scope.Log(fmt.Sprintf("failed to update stats: %v", err))
 		}
 	}
+
+	// Update the per-realm stats.
+	sql := `
+		INSERT INTO realm_stats(date, realm_id, codes_issued)
+			VALUES ($1, $2, 1)
+		ON CONFLICT (date, realm_id) DO UPDATE
+			SET codes_issued = realm_stats.codes_issued + 1
+	`
+
+	if err := scope.DB().Exec(sql, v.CreatedAt.Truncate(24*time.Hour), v.RealmID).Error; err != nil {
+		scope.Log(fmt.Sprintf("failed to update stats: %v", err))
+	}
+}
+
+// claim checks the VerificationCode, marks as claimed, and updates the
+// per-realm stats. If any of that fails, an error code is returned.
+func (v *VerificationCode) claim(db *Database, tx *gorm.DB, verCode string, realmID uint) error {
+	if expired, err := db.IsCodeExpired(v, verCode); err != nil {
+		db.logger.Debugw("error checking code expiration", "error", err)
```

Comment:
I think we should drop this log message (and the dependency on `db`) and return the error instead. That'll make it bubble as an internalerror instead of silently swallowing right now.

Reply:
you got it.